src/main/java/uk/nhs/hee/tis/trainee/forms/service/FormRelocateService.java

```diff
@@ -0,0 +1,186 @@
+            relocateFormInS3(formType, formId, targetTrainee, sourceTrainee);
+          }
+        } catch (Exception ex) {
+          log.error("Fail to roll back: " + ex);
+        }
+        throw new ApplicationException("Fail to relocate FormR to target trainee: " + e.toString());
+      }
+    }
+  }
+
+  private AbstractForm getMoveFormInfoInDb(String formId) {
+    try {
+      Optional<FormRPartA> optionalFormRPartA =
+          formRPartARepository.findById(UUID.fromString(formId));
+      Optional<FormRPartB> optionalFormRPartB =
+          formRPartBRepository.findById(UUID.fromString(formId));
+
+      if (optionalFormRPartA.isPresent()) {
+        return optionalFormRPartA.get();
+      }
+      else if (optionalFormRPartB.isPresent()) {
+        return optionalFormRPartB.get();
+      }
+      return null;
+    } catch (Exception e) {
+      log.error("Fail to get FormR with ID " + formId + ": " + e);
+      throw new ApplicationException("Fail to get FormR with ID " + formId + ": " + e.toString());
+    }
+  }
+
+  private void updateTargetTraineeInDb(AbstractForm abstractForm, String targetTrainee) {
+    abstractForm.setTraineeTisId(targetTrainee);
+
+    if (abstractForm instanceof FormRPartA formRPartA) {
+      formRPartARepository.save(formRPartA);
+    } else if (abstractForm instanceof FormRPartB formRPartB) {
+      formRPartBRepository.save(formRPartB);
+    }
+    log.info("Form R with ID " + abstractForm.getId()
+        + " moved under " + targetTrainee + " in DB ");
+  }
+
+  private void relocateFormInS3(
+      String formType, String formId, String sourceTrainee, String targetTrainee)
+      throws IOException {
+
+    String sourceKey = sourceTrainee + "/forms/" + formType + "/" + formId + ".json";
+    String targetKey = targetTrainee + "/forms/" + formType + "/" + formId + ".json";
+
+    log.info("sourceKey : " + sourceKey);
+    log.info("targetKey : " + targetKey);
+
+    // Get Form R from S3
+    S3Object object = amazonS3.getObject(bucketName, sourceKey);
+
+    if (object != null) {
+      // Set target trainee Id to S3 metadata
+      ObjectMetadata metadata = object.getObjectMetadata();
+      metadata.addUserMetadata("traineeid", targetTrainee);
+
+      // Set target trainee Id to S3 content
+      S3ObjectInputStream content = object.getObjectContent();
+      ObjectMapper mapper = new ObjectMapper();
+      Map<String, Object> formr = mapper.readValue(content, Map.class);
+      formr.put("traineeTisId", targetTrainee);
+
+      // Save the form under target trainee
+      final PutObjectRequest request = new PutObjectRequest(
+          bucketName,
+          targetKey,
+          new ByteArrayInputStream(objectMapper.writeValueAsBytes(formr)),
+          metadata);
+      amazonS3.putObject(request);
+      log.info("Form R in S3 relocated from " + sourceKey + " to " + targetKey);
+
+      // Delete Form R from source trainee
+      amazonS3.deleteObject(bucketName, sourceKey);
```

Comment:
On the ticket it says 'The old form is removed from the NDW (existing partial delete to keep a history is fine)' - this code looks like a real (not partial) delete. Is that right? I think it might need to be changed if so.

Reply:
Thank you @ReubenRobertsHEE for the comment and discussion. 🌮 
It's a good shout. As you have mentioned, the ticket is promoting partial delete to keep history, while earlier conversation in the channel suggested deleting the forms directly from S3 to avoid NDW clashes: [chat link](https://hee-nhs-tis.slack.com/archives/G0135LS4JVA/p1698248857703269?thread_ts=1698246122.129819&cid=G0135LS4JVA)
@Judge40 do you have it in mind whether we should partial or hard delete this in S3? If partial delete is conducted, does it mean we should halt NDW export and remove the messages from the SQS queue in Lambda to avoid clash?

Reply:
It's a question for Naz to answer really, since it affects what users will see (or not see) on downstream reports.

Depending on the decision:

 - Hard delete: no issue here since deletes don't send an S3 notification to the exporter. Possibly risk if that ever changed.
 - Soft delete: generating a new form ID seems like the sensible option to avoid race conditions and/or messing with queues. Having multiple forms with the same uuid is bound to cause us an issue at some point and just feels wrong.

Another thing we might want to consider is whether we should be setting a "delete reason" on soft deleted forms. In our scenarios, that would be with a value of either
1. DELETED
2. MOVED

Do we also need a record of where a form was moved to, or can we rely on logs if we ever need to answer that?

Reply:
As confirmed by Naz, we will hard delete the old formR record when the form is moved.